src/Sylius/Component/User/Security/Provider/AbstractUniqueTokenProvider.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\User\Security\Provider;
+
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Sylius\Component\User\Security\Generator\GeneratorInterface;
+
+/**
+ * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
+ * @author Jan Góralski <jan.goralski@lakion.com>
+ */
+abstract class AbstractUniqueTokenProvider implements UniqueTokenProviderInterface
+{
+    /**
+     * @var RepositoryInterface
+     */
+    protected $userRepository;
+
+    /**
+     * @var GeneratorInterface
+     */
+    protected $generator;
+
+    /**
+     * @var int
+     */
+    protected $tokenLength;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function generateUniqueToken()
```

Comment:
We make may it `final` as overriding it destroys all the abstraction this class gives.